engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ModuleIndexing.scala

```diff
@@ -0,0 +1,48 @@
+package org.enso.interpreter.instrument.execution
+
+import org.enso.interpreter.runtime.Module
+
+import java.util.concurrent.{ConcurrentHashMap, ConcurrentMap};
+
+/** Modules need to be indexed for building suggestions DB. */
+final class ModuleIndexing {
+
+  object IndexState extends Enumeration {
+    val NotIndexed, NeedsIndexing, Indexed = Value
+  }
+
+  private val modules: ConcurrentMap[Module, IndexState.Value] =
```

Comment:
The code is now rewritten into a new `ModuleIndexing` Scala class that uses `ConcurrentHashMap`. Good encapsulation, but I still miss the answer to the [fundamental question](https://github.com/enso-org/enso/pull/9639#discussion_r1556904725): What is the **invariant** to be consistent with?